src/RpcClient.ts

```diff
@@ -424,6 +435,144 @@ export class RpcClient {
+    payer: PublicKey,
+    lookupTables: AddressLookupTableAccount[]
+  ): Promise<number | null> {
+    const testInstructions = [
+      ComputeBudgetProgram.setComputeUnitLimit({ units: 1_400_000 }),
+      ...instructions,
+    ];
+
+    const testTransaction = new VersionedTransaction(
+      new TransactionMessage({
+        instructions: testInstructions,
+        payerKey: payer,
+        recentBlockhash: (await this.connection.getLatestBlockhash()).blockhash,
+      }).compileToV0Message(lookupTables)
+    );
+
+    const rpcResponse = await this.connection.simulateTransaction(testTransaction, {
+      replaceRecentBlockhash: true,
+      sigVerify: false,
+    });
+
+    if (rpcResponse.value.err) {
+      console.error(`Simulation error: ${rpcResponse.value.err}`);
+      return null;
+    }
+
+    return rpcResponse.value.unitsConsumed || null;
+  }
+
+  /**
+   * Poll a transaction to confirm whether it has been confirmed
+   * @param {TransactionSignature} txtSig - The transaction signature
+   * @returns {Promise<TransactionSignature>} - The confirmed transaction signature
+  */
+  async pollTransactionConfirmation(txtSig: TransactionSignature): Promise<TransactionSignature> {
+    // 30 second timeout
+    const timeout = 30000;
+    // 5 second retry interval
+    const interval = 5000;
+    let elapsed = 0;
+
+    return new Promise<TransactionSignature>((resolve, reject) => {
+      const intervalId = setInterval(async () => {
+        elapsed += interval;
+
+        if (elapsed >= timeout) {
+          clearInterval(intervalId);
+          reject(new Error(`Transaction ${txtSig}'s confirmation timed out`));
+        }
+
+        const status = await this.connection.getSignatureStatus(txtSig);
+
+        if (status?.value?.confirmationStatus === "finalized") {
+          clearInterval(intervalId);
+          resolve(txtSig);
+        }
+      }, interval);
+    });
+  }
+
+  /**
+   * Build and send an optimized transaction, and handle its confirmation status
+   * @param {TransactionInstruction[]} instructions - The transaction instructions
+   * @param {Keypair} fromKeypair - The sender's keypair
+   * @param {boolean} skipPreflightChecks - Whether the transaction should skip preflight checks. Defaults to `true`
+   * @param {number} maxRetries - The maximum number of times to retry sending the transaction to the leader. Defaults to 0
+   * @returns {Promise<TransactionSignature>} - The transaction signature
+  */
+  async sendSmartTransaction(
+    instructions: TransactionInstruction[],
+    fromKeypair: Keypair,
+    skipPreflightChecks: boolean = true,
+    maxRetries: number = 0,
+  ): Promise<TransactionSignature> {
+    try {
+      const pubKey = fromKeypair.publicKey;
+
+      // Build the initial transaction to estimate the priority fee
+      const transaction = new Transaction().add(...instructions);
+      transaction.recentBlockhash = (await this.connection.getLatestBlockhash()).blockhash;
```

Comment:
you can send all of the promises out at the same time to reduce latency, instead of running each one and waiting for it to return